buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -352,7 +595,527 @@ static String decodeString(ByteBuffer src, Charset charset) {
+                        decoder.decode(heapBuffer.internalNioBuffer(heapBuffer.readerIndex(), length));
+                    } finally {
+                        heapBuffer.release();
+                    }
+                }
+                return true;
+            } catch (CharacterCodingException ignore) {
+                return false;
+            }
+        }
+    }
+
+    /**
+     * Aborts on a byte which is not a valid ASCII character.
+     */
+    private static final ByteBufProcessor FIND_NON_ASCII = new ByteBufProcessor() {
+        @Override
+        public boolean process(byte value) {
+            return value >= 0;
+        }
+    };
+
+    /**
+     * Returns {@code true} if the specified {@link ByteBuf} starting at {@code index} with {@code length} is valid
+     * ASCII text, otherwise return {@code false}.
+     *
+     * @param buf    The given {@link ByteBuf}.
+     * @param index  The start index of the specified buffer.
+     * @param length The length of the specified buffer.
+     */
+    private static boolean isAscii(ByteBuf buf, int index, int length) {
+        return buf.forEachByte(index, length, FIND_NON_ASCII) == -1;
+    }
+
+    /**
+     * Returns {@code true} if the specified {@link ByteBuf} starting at {@code index} with {@code length} is valid
+     * UTF8 text, otherwise return {@code false}.
+     *
+     * @param buf The given {@link ByteBuf}.
+     * @param index The start index of the specified buffer.
+     * @param length The length of the specified buffer.
+     *
+     * @see
+     * <a href=http://www.ietf.org/rfc/rfc3629.txt>UTF-8 Definition</a>
+     *
+     * <pre>
+     * 1. Bytes format of UTF-8
+     *
+     * The table below summarizes the format of these different octet types.
+     * The letter x indicates bits available for encoding bits of the character number.
+     *
+     * Char. number range  |        UTF-8 octet sequence
+     *    (hexadecimal)    |              (binary)
+     * --------------------+---------------------------------------------
+     * 0000 0000-0000 007F | 0xxxxxxx
+     * 0000 0080-0000 07FF | 110xxxxx 10xxxxxx
+     * 0000 0800-0000 FFFF | 1110xxxx 10xxxxxx 10xxxxxx
+     * 0001 0000-0010 FFFF | 11110xxx 10xxxxxx 10xxxxxx 10xxxxxx
+     * </pre>
+     *
+     * <pre>
+     * 2. Syntax of UTF-8 Byte Sequences
+     *
+     * UTF8-octets = *( UTF8-char )
+     * UTF8-char   = UTF8-1 / UTF8-2 / UTF8-3 / UTF8-4
+     * UTF8-1      = %x00-7F
+     * UTF8-2      = %xC2-DF UTF8-tail
+     * UTF8-3      = %xE0 %xA0-BF UTF8-tail /
+     *               %xE1-EC 2( UTF8-tail ) /
+     *               %xED %x80-9F UTF8-tail /
+     *               %xEE-EF 2( UTF8-tail )
+     * UTF8-4      = %xF0 %x90-BF 2( UTF8-tail ) /
+     *               %xF1-F3 3( UTF8-tail ) /
+     *               %xF4 %x80-8F 2( UTF8-tail )
+     * UTF8-tail   = %x80-BF
+     * </pre>
+     */
+    private static boolean isUtf8(ByteBuf buf, int index, int length) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "isUtf8" is 36 which is greater than 25 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)